src/Streamly/Internal/Data/Parser/Types.hs

```diff
@@ -181,7 +252,220 @@ splitWith func (Parser stepL initialL extractL)
+-- quadratic time complexity, because each composition adds a new branch that
+-- each subsequent alternative's input element has to go through, therefore, it
+-- cannot scale to a large number of compositions
+--
+-- | Sequential alternative. Apply the input to the first parser and return the
+-- result if the parser succeeds. If the first parser fails then backtrack and
+-- apply the same input to the second parser and return the result.
+--
+-- Note: This implementation is not lazy in the second argument. The following
+-- will fail:
+--
+-- >>> S.parse (PR.satisfy (> 0) `PR.alt` undefined) $ S.fromList [1..10]
+--
+-- /Internal/
+--
+{-# INLINE alt #-}
+alt :: Monad m => Parser m x a -> Parser m x a -> Parser m x a
+alt (Parser stepL initialL extractL) (Parser stepR initialR extractR) =
+    Parser step initial extract
+
+    where
+
+    initial = AltParseL 0 <$> initialL
+
+    -- Once a parser yields at least one value it cannot fail.  This
+    -- restriction helps us make backtracking more efficient, as we do not need
+    -- to keep the consumed items buffered after a yield. Note that we do not
+    -- enforce this and if a misbehaving parser does not honor this then we can
+    -- get unexpected results.
+    step (AltParseL cnt st) a = do
+        r <- stepL st a
+        case r of
+            Yield n s -> return $ Yield n (AltParseL 0 s)
+            Skip n s -> do
+                assert (cnt + 1 - n >= 0) (return ())
+                return $ Skip n (AltParseL (cnt + 1 - n) s)
+            Stop n b -> return $ Stop n b
+            Error _ -> do
+                rR <- initialR
+                return $ Skip (cnt + 1) (AltParseR rR)
+
+    step (AltParseR st) a = do
+        r <- stepR st a
+        return $ case r of
+            Yield n s -> Yield n (AltParseR s)
+            Skip n s -> Skip n (AltParseR s)
+            Stop n b -> Stop n b
+            Error err -> Error err
+
+    extract (AltParseR sR) = extractR sR
+    extract (AltParseL _ sL) = extractL sL
+
+-- | See documentation of 'Streamly.Internal.Data.Parser.many'.
+--
+-- /Internal/
+--
+{-# INLINE splitMany #-}
+splitMany :: MonadCatch m => Fold m b c -> Parser m a b -> Parser m a c
+splitMany (Fold fstep finitial fextract) (Parser step1 initial1 extract1) =
+    Parser step initial extract
+
+    where
+
+    initial = do
+        ps <- initial1 -- parse state
+        fs <- finitial -- fold state
+        pure (Tuple3' ps 0 fs)
+
+    {-# INLINE step #-}
+    step (Tuple3' st cnt fs) a = do
+        r <- step1 st a
+        let cnt1 = cnt + 1
+        case r of
+            Yield _ s -> return $ Skip 0 (Tuple3' s cnt1 fs)
+            Skip n s -> do
+                assert (cnt1 - n >= 0) (return ())
+                return $ Skip n (Tuple3' s (cnt1 - n) fs)
+            Stop n b -> do
+                s <- initial1
+                fs1 <- fstep fs b
+                -- XXX we need to yield and backtrack here
```

Comment:
Modifying the state will probably let you do that.

Reply:
Yes, we will fix it in a separate change.